src/node_snapshotable.cc

```diff
@@ -166,175 +177,182 @@ void SnapshotBuilder::InitializeIsolateParams(const SnapshotData* data,
+        main_instance->isolate_data()->Serialize(&creator);
+
+    // The default context with only things created by V8.
+    Local<Context> default_context = Context::New(isolate);
+
+    // The Node.js-specific context with primodials, can be used by workers
+    // TODO(joyeecheung): investigate if this can be used by vm contexts
+    // without breaking compatibility.
+    Local<Context> base_context = NewContext(isolate);
+    if (base_context.IsEmpty()) {
+      return INTERNAL_ERROR;
+    }
+
+    Local<Context> main_context = NewContext(isolate);
+    if (main_context.IsEmpty()) {
+      return INTERNAL_ERROR;
+    }
+    // Initialize the main instance context.
+    {
+      Context::Scope context_scope(main_context);
+
+      // Create the environment.
+      env = new Environment(main_instance->isolate_data(),
+                            main_context,
+                            args,
+                            exec_args,
+                            nullptr,
+                            node::EnvironmentFlags::kDefaultFlags,
+                            {});
+
+      // Run scripts in lib/internal/bootstrap/
+      if (env->RunBootstrapping().IsEmpty()) {
+        return INTERNAL_ERROR;
+      }
+      // If --build-snapshot is true, lib/internal/main/mksnapshot.js would be
+      // loaded via LoadEnvironment() to execute process.argv[1] as the entry
+      // point (we currently only support this kind of entry point, but we
+      // could also explore snapshotting other kinds of execution modes
+      // in the future).
+      if (per_process::cli_options->build_snapshot) {
 #if HAVE_INSPECTOR
-          env->InitializeInspector({});
+        // TODO(joyeecheung): move this before RunBootstrapping().
+        env->InitializeInspector({});
 #endif
-          // TODO(joyeecheung): we could use the result for something special,
-          // like setting up initializers that should be invoked at snapshot
-          // dehydration.
-          MaybeLocal<Value> result =
-              LoadEnvironment(env, StartExecutionCallback{});
-          if (bootstrapCatch.HasCaught()) {
-            // TODO(joyeecheung): fail by exiting with a non-zero exit code.
-            PrintCaughtException(isolate, main_context, bootstrapCatch);
-            abort();
-          }
-          result.ToLocalChecked();
-          // FIXME(joyeecheung): right now running the loop in the snapshot
-          // builder seems to introduces inconsistencies in JS land that need to
-          // be synchronized again after snapshot restoration.
-          int exit_code = SpinEventLoop(env).FromMaybe(1);
-          CHECK_EQ(exit_code, 0);
-          if (bootstrapCatch.HasCaught()) {
-            // TODO(joyeecheung): fail by exiting with a non-zero exit code.
-            PrintCaughtException(isolate, main_context, bootstrapCatch);
-            abort();
-          }
+        if (LoadEnvironment(env, StartExecutionCallback{}).IsEmpty()) {
+          return 1;
```

Comment:
Will this path be taken only if an exception is thrown from the user script?

Do we want to differentiate such an error from the other kinds of errors that have the same exit code, like incorrect number of arguments in https://github.com/nodejs/node/blob/e80613c70d6c1efc6c6bbef85c95569d13ffdef4/tools/snapshot/node_mksnapshot.cc#L60-L65?

Should we turn the `1` into a symbol like `INTERNAL_ERROR` and reuse it in `tools/snapshot/node_mksnapshot.cc` in case we don't want to differentiate between such kinds of errors?

Reply:
12 is one of the exit code that we already use to signal internal error (another is 9, though I have no idea how they differ)

Reply:
Also, mksnapshot is already an internal tool, so I guess its exit code is less relevant, but if we want to make it follow whatever the actual user-facing executable uses, that should've been 12 or 9.

Reply:
Ah, are you referring to the exit codes documented in https://nodejs.org/api/process.html#exit-codes?
Neither 9 nor 12 seems fitting for "internal error" because those are for "Invalid Argument" and "Invalid Debug Argument" respectively.

The 1 exit code selection for the LoadEnvironment failure LGTM though.

Reply:
I've added 14 as an exit code for snapshot failures, and left a TODO for sharing the exit code constants across the code base.